### ISSUE TITLE
reuse globalThis logic for JSX namespaced custom elements

### ISSUE THREAD

**Issue by jeremy-coleman** [Microsoft/TypeScript#30658]
## Suggestion

It seems likely the compiler logic used for globalThis could be reused to support automated global types for custom elements , which is especially annoying and difficult to deal with in jsx

## Use Cases

loading-spinner.ts
```ts

 class LoadingSpinner extends HTMLElement {
  constructor() {
    super();
  this.innerHTML =  `<div class="loading-spinnner">` 
   }
//omitting code for brevity
}

customElements.define('loading-spinner', LoadingSpinner);
```

loading-spinner.d.ts
```ts
interface LoadingSpinner extends JSX.HTMLAttributes {}

declare namespace JSX {
  interface IntrinsicElements {
    'loading-spinner': LoadingSpinner;
  }
}
```

```ts
import h from Reactish
let MyComponent = () => {
return (
 <loading-spinner/> //<--so difficult not to get errors here
)
}

**Comment by DanielRosenwasser** [comment]
I don't think so, but it sounds like you want an easier way to apply global augmentations for registration patterns which are necessary for custom elements, which is a reasonable request.

**Comment by jeremy-coleman** [comment]
Ah, was hoping it could be implemented easily by treating the customElement  registry similar to a global var, but yes you summed it up perfectly. There is really a need beyond just custom elements , but i think that is the best target to aim for currently. On a related note, the stencil compiler generates types and writes them to disk on the fly , which may be a useful resource if you decide to pursue this
